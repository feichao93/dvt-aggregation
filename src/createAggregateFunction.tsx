// @ts-ignore
import _parser from './grammar.pegjs'

// @ts-ignore
const version = DVT_AGGREGATION_VERSION

const parser = _parser as {
  parse(expression: string): Expression
}

type AggrFuncName = 'MAX' | 'MIN' | 'SUM' | 'AVG' | 'COUNT'
type OP = '+' | '-' | '*' | '/'

type Expression = NumberLiteral | Field | ArithmeticExpression | AggrFuncCallExpression
type NumberLiteral = { type: 'n'; n: number }
type Field = { type: 'field'; code: string }
type ArithmeticExpression = { type: 'arithmetic'; ops: OP[]; terms: Expression[] }
type AggrFuncCallExpression = { type: 'aggr-func-call'; funcName: AggrFuncName; subExpr: Expression }

function collectCodeWithDepFlag(rootExpr: Expression) {
  const result: Array<{ code: string; isDep: boolean }> = []
  dfs(rootExpr, false)
  return result

  function dfs(expr: Expression, inAggrFuncCall: boolean) {
    if (expr.type === 'arithmetic') {
      for (const term of expr.terms) {
        dfs(term, false)
      }
    } else if (expr.type === 'field') {
      const index = result.findIndex(({ code }) => code === expr.code)
      if (index === -1) {
        result.push({ code: expr.code, isDep: !inAggrFuncCall })
      } else {
        if (!inAggrFuncCall) {
          result[index].isDep = true
        }
      }
    } else if (expr.type === 'aggr-func-call') {
      dfs(expr.subExpr, true)
    }
  }
}

// todo 不允许嵌套聚合调用
// todo 更加严格的输入检查
function convertItems(items: Item[]): Item[] {
  let safeCodeCount = 0
  const genSafeCode = () => `safe_code_${safeCodeCount++}`

  return items.flatMap(dfs)

  function dfs(item: Item): Item[] {
    const { code, deps, expr } = item
    const result: Item[] = []
    if (expr.type === 'aggr-func-call') {
      if (expr.funcName === 'AVG') {
        const numerator = genSafeCode() // 分子
        const denominator = genSafeCode() // 分母
        result.push(
          ...dfs({
            code: numerator,
            deps: [],
            expr: {
              type: 'aggr-func-call',
              funcName: 'SUM',
              subExpr: expr.subExpr,
            },
          }),
          ...dfs({
            code: denominator,
            deps: [],
            expr: {
              type: 'aggr-func-call',
              funcName: 'COUNT',
              subExpr: expr.subExpr,
            },
          }),
          {
            code,
            deps: [numerator, denominator],
            expr: {
              type: 'arithmetic',
              terms: [
                { type: 'field', code: numerator },
                { type: 'field', code: denominator },
              ],
              ops: ['/'],
            },
          },
        )
      } else if (expr.subExpr.type === 'arithmetic') {
        const safeCode = genSafeCode()
        result.push(
          ...dfs({
            code,
            deps: [],
            expr: {
              type: 'aggr-func-call',
              funcName: expr.funcName,
              subExpr: { type: 'field', code: safeCode },
            },
          }),
          ...dfs({
            code: safeCode,
            deps: [],
            expr: expr.subExpr,
          }),
        )
      } else {
        result.push(item)
      }
    } else if (expr.type === 'arithmetic') {
      const transformedTerms: Expression[] = []
      const appendDeps: string[] = []
      for (const term of expr.terms) {
        if (term.type === 'aggr-func-call') {
          const safeCode = genSafeCode()
          result.push(
            ...dfs({
              code: safeCode,
              deps: [],
              expr: {
                type: 'aggr-func-call',
                funcName: term.funcName,
                subExpr: term.subExpr,
              },
            }),
          )
          transformedTerms.push({ type: 'field', code: safeCode })
          appendDeps.push(safeCode)
        } else {
          transformedTerms.push(term)
        }
      }
      result.push({
        code,
        expr: { ...expr, terms: transformedTerms },
        deps: mergeDeps(deps, appendDeps),
      })
    } else {
      result.push(item)
    }

    return result
  }
}

type Item = { code: string; expr: Expression; deps: string[] }

export interface IndicatorConfig {
  name: string
  code: string
  expression?: string
}

function parseIndicators(indicators: IndicatorConfig[]): Item[] {
  const allIndCodes = new Set(indicators.map(ind => ind.code))
  return indicators.map(ind => {
    if (ind.expression == null) {
      throw new Error(`${ind.code}:${ind.name} 缺少聚合表示式`)
    }
    let expr: Expression
    try {
      expr = parser.parse(ind.expression)
    } catch (e) {
      console.warn(`解析 ${ind.code}:${ind.name} 的表达式失败，请仔细检查下列表示是否正确` + `\n\t${ind.expression}`)
      throw e
    }
    const codesWithDepFlag = collectCodeWithDepFlag(expr)
    codesWithDepFlag.forEach(({ code }) => {
      if (!allIndCodes.has(code)) {
        throw new Error(`${code} 不是合法的指标 code`)
      }
    })
    return {
      code: ind.code,
      expr,
      deps: codesWithDepFlag.filter(({ isDep }) => isDep).map(({ code }) => code),
    }
  })
}

function emit(indicators: IndicatorConfig[], items: Item[]) {
  const sumItems = items.filter(({ expr }) => expr.type === 'aggr-func-call' && expr.funcName === 'SUM')
  const countItems = items.filter(({ expr }) => expr.type === 'aggr-func-call' && expr.funcName === 'COUNT')
  const maxItems = items.filter(({ expr }) => expr.type === 'aggr-func-call' && expr.funcName === 'MAX')
  const minItems = items.filter(({ expr }) => expr.type === 'aggr-func-call' && expr.funcName === 'MIN')

  const sumPreparation = sumItems
    .map(({ code }) => {
      return `${t.field('r', code)} = 0;`
    })
    .join('\n')

  const countPreparation = countItems
    .map(({ code }) => {
      return `${t.field('r', code)} = 0;`
    })
    .join('\n')

  const maxPreparation = maxItems
    .map(({ code }) => {
      return `${t.field('r', code)} = -Infinity;`
    })
    .join('\n')

  const minPreparation = minItems
    .map(({ code }) => {
      return `${t.field('r', code)} = Infinity;`
    })
    .join('\n')

  const countCalculation = countItems
    .map(({ code, expr }) => {
      if (expr.type === 'aggr-func-call') {
        return `${t.field('r', code)} += ${t.field('d', code, '1')};`
      } else {
        throw new Error()
      }
    })
    .join('\n')

  const sumCalculation = sumItems
    .map(({ code, expr }) => {
      const subExpr = (expr as AggrFuncCallExpression).subExpr
      return `${t.field('r', code)} += ${t.makeJSExpression(subExpr, 'd')} || 0;`
    })
    .join('\n')

  const maxCalculation = maxItems
    .map(({ code, expr }) => {
      const subExpr = (expr as AggrFuncCallExpression).subExpr
      return `${t.field('r', code)} = Math.max(${t.field('r', code)}, ${t.makeJSExpression(subExpr, 'd')});`
    })
    .join('\n')

  const minCalculation = minItems
    .map(({ code, expr }) => {
      const subExpr = (expr as AggrFuncCallExpression).subExpr
      return `${t.field('r', code)} = Math.min(${t.field('r', code)}, ${t.makeJSExpression(subExpr, 'd')});`
    })
    .join('\n')

  const derivedCalculation = items
    .filter(({ code, expr }) => {
      return expr.type !== 'aggr-func-call'
    })
    .map(({ code, expr }) => {
      return `${t.field('r', code)} = ${t.makeJSExpression(expr, 'r')};`
    })
    .join('\n')

  // 这里需要使用 [...].join('\n')，不能直接使用模板字符串，否则 rollup 在编译的时候会把代码破坏掉
  return [
    t.__(`/* aggregate function generated by dvt-aggregation@${version}`),
    t.____(...indicators.map(ind => `${ind.name}: ${ind.code} = ${ind.expression}`)),
    t.__('*/'),
    '',

    t.__('let r = {};'),
    '',
    t.__(countPreparation, sumPreparation, maxPreparation, minPreparation),
    '',
    t.__('for (const d of slice) {'),
    t.____(sumCalculation, countCalculation, maxCalculation, minCalculation),
    t.__('}'),
    '',
    t.__(derivedCalculation),
    '',
    t.__('return r;'),
  ].join('\n')
}

const t = {
  __(...ss: string[]) {
    return t.indent(1, ...ss)
  },
  ____(...ss: string[]) {
    return t.indent(2, ...ss)
  },
  indent(n: number, ...ss: string[]) {
    return ss
      .join('\n')
      .split('\n')
      .filter(s => !s.match(/^\s*$/))
      .map(line => '  '.repeat(n).concat(line.trim()))
      .join('\n')
  },

  field(objectId: string, field: string, defaultValue?: string) {
    if (defaultValue == null) {
      return `${objectId}['${field}']`
    }
    return `(${objectId}['${field}'] || ${defaultValue})`
  },

  makeJSExpression(rootExpr: Expression, refObjectId: string) {
    return dfs(rootExpr)

    function dfs(expr: Expression): string {
      if (expr.type === 'arithmetic') {
        const termsJSExpr = expr.terms.map(dfs)
        const opJSExpr = expr.ops
        const result = [termsJSExpr[0]]
        for (let i = 0; i < opJSExpr.length; i++) {
          result.push(opJSExpr[i], termsJSExpr[i + 1])
        }
        return `(${result.join(' ')})`
      } else if (expr.type === 'field') {
        return t.field(refObjectId, expr.code)
      } else if (expr.type === 'n') {
        return String(expr.n)
      } else {
        throw new Error('当前暂不支持该运算过程')
      }
    }
  },
}

function topologicalSort<T extends { code: string; deps: string[] }>(array: T[]): T[] {
  const result: T[] = []

  const size = array.length

  const edgesMap = new Map<string, number[]>()
  for (let i = 0; i < size; i++) {
    for (const dep of array[i].deps) {
      if (!edgesMap.has(dep)) {
        edgesMap.set(dep, [])
      }
      edgesMap.get(dep).push(i)
    }
  }

  const indegrees = array.map(({ deps }) => deps.length)
  const used = new Array(size).fill(false)

  while (true) {
    let i = 0
    for (; i < size; i++) {
      if (indegrees[i] === 0 && !used[i]) {
        break
      }
    }
    if (i === size) {
      break
    }
    used[i] = true
    const item = array[i]
    result.push(item)
    if (edgesMap.has(item.code)) {
      for (const index of edgesMap.get(item.code)) {
        indegrees[index]--
      }
    }
  }
  if (result.length !== size) {
    const firstFailCode = array[used.findIndex(u => !u)].code
    const circle = findCircularDependency(firstFailCode, array)
    throw new Error('指标计算之间存在相互依赖关系\n\t' + circle.join(' -> ') + '\n（说明：A -> B 表示A的计算依赖于B）')
  }

  return result
}

function findCircularDependency(startCode: string, array: Array<{ code: string; deps: string[] }>) {
  const graph = new Map(array.map(({ code, deps }) => [code, deps]))
  const circle: string[] = []
  dfs(startCode)
  return circle

  function dfs(currentCode: string) {
    if (circle.length > 0 && circle[0] === currentCode) {
      circle.push(currentCode)
      return true
    }
    circle.push(currentCode)
    if (graph.has(currentCode)) {
      for (const nextCode of graph.get(currentCode)) {
        if (dfs(nextCode)) {
          return true
        }
      }
    }
    circle.pop()
    return false
  }
}

function mergeDeps(deps: string[], appendDeps: string[]) {
  const set = new Set(deps)
  const result = deps.slice()
  for (const d of appendDeps) {
    if (!set.has(d)) {
      set.add(d)
      result.push(d)
    }
  }
  return result
}

export default function createAggregateFunction(indicators: IndicatorConfig[]): (slice: any[]) => any {
  const template = emit(indicators, topologicalSort(convertItems(parseIndicators(indicators))))
  return new Function('slice', template) as any
}
