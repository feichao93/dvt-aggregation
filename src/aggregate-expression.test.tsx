import createAggregateFunction, { IndicatorConfig } from './createAggregateFunction'

const testData = [
  { income: 20, target: 80 },
  { income: 10, target: 20 },
  { income: 30, target: 50 },
  { income: 40, target: 90 },
  // sum:  { income: 100, target: 240 }
]

describe('SUM 聚合表达式', () => {
  const indicators: IndicatorConfig[] = [
    { code: 'income', name: '收入', expression: 'SUM(income)' },
    { code: 'target', name: '目标', expression: 'SUM(target)' },
  ]
  const aggregate = createAggregateFunction(indicators)

  test('编译结果', () => {
    expect(aggregate.toString()).toMatchSnapshot()
  })

  test('运行结果', () => {
    const actual = aggregate(testData)
    expect(actual.income).toBeCloseTo(100)
    expect(actual.target).toBeCloseTo(240)
  })
})

describe('简单的计算表达式', () => {
  const indicators: IndicatorConfig[] = [
    { code: 'income', name: '收入', expression: 'SUM(income)' },
    { code: 'target', name: '目标', expression: 'SUM(target)' },
    { code: 'act_rate', name: '达成率', expression: 'income / target' },
  ]
  const aggregate = createAggregateFunction(indicators)

  test('编译结果', () => {
    expect(aggregate.toString()).toMatchSnapshot()
  })

  test('运行结果', () => {
    const actual = aggregate(testData)
    expect(actual.act_rate).toBeCloseTo(100 / 240)
  })
})

describe('AVG/MAX/MIN 聚合表达式', () => {
  const indicators: IndicatorConfig[] = [
    { code: 'income', name: 'income', expression: 'SUM(income)' },
    { code: 'target', name: 'target', expression: 'SUM(target)' },
    { code: 'max_income', name: '最大收入', expression: 'MAX(income)' },
    { code: 'min_target', name: '最小目标', expression: 'MIN(target)' },
    { code: 'avg_income', name: '平均收入', expression: 'AVG(income)' },
  ]
  const aggregate = createAggregateFunction(indicators)

  test('编译结果', () => {
    expect(aggregate.toString()).toMatchSnapshot()
  })

  test('运行结果', () => {
    const actual = aggregate(testData)
    expect(actual.max_income).toBeCloseTo(40)
    expect(actual.min_target).toBeCloseTo(20)
    expect(actual.avg_income).toBeCloseTo(25)
  })
})

describe('复杂的计算表达式', () => {
  const indicators: IndicatorConfig[] = [
    { code: 'income', name: '收入', expression: 'SUM(income)' },
    { code: 'target', name: '目标', expression: 'SUM(target)' },
    { code: 'code_1', name: 'code_1', expression: '(target - income) / SUM(income)' },
    { code: 'code_2', name: 'code_2', expression: 'MAX(target) - MIN(income)' },
    { code: 'code_3', name: 'code_3', expression: 'MIN(target) - MAX(income)' },
    { code: 'code_4', name: 'code_4', expression: 'code_2 + code_3' },
  ]
  const aggregate = createAggregateFunction(indicators)

  test('编译结果', () => {
    expect(aggregate.toString()).toMatchSnapshot()
  })

  test('运行结果', () => {
    const actual = aggregate(testData)
    expect(actual.code_1).toBeCloseTo((240 - 100) / 100)
    expect(actual.code_2).toBeCloseTo(90 - 10)
    expect(actual.code_3).toBeCloseTo(20 - 40)
    expect(actual.code_4).toBeCloseTo(actual.code_2 + actual.code_3)
  })
})
