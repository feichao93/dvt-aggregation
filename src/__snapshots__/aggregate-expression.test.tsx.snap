// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AVG/MAX/MIN 聚合表达式 编译结果 1`] = `
"function anonymous(slice
) {
  /* aggregate function generated by dvt-aggregation@test
    income: income = SUM(income)
    target: target = SUM(target)
    最大收入: max_income = MAX(income)
    最小目标: min_target = MIN(target)
    平均收入: avg_income = AVG(income)
  */

  let r = {};

  r['safe_code_1'] = 0;
  r['income'] = 0;
  r['target'] = 0;
  r['safe_code_0'] = 0;
  r['max_income'] = -Infinity;
  r['min_target'] = Infinity;

  for (const d of slice) {
    r['income'] += d['income'] || 0;
    r['target'] += d['target'] || 0;
    r['safe_code_0'] += d['income'] || 0;
    r['safe_code_1'] += (d['safe_code_1'] || 1);
    r['max_income'] = Math.max(r['max_income'], d['income']);
    r['min_target'] = Math.min(r['min_target'], d['target']);
  }

  r['avg_income'] = (r['safe_code_0'] / r['safe_code_1']);

  return r;
}"
`;

exports[`SUM 聚合表达式 编译结果 1`] = `
"function anonymous(slice
) {
  /* aggregate function generated by dvt-aggregation@test
    收入: income = SUM(income)
    目标: target = SUM(target)
  */

  let r = {};

  r['income'] = 0;
  r['target'] = 0;

  for (const d of slice) {
    r['income'] += d['income'] || 0;
    r['target'] += d['target'] || 0;
  }



  return r;
}"
`;

exports[`复杂的计算表达式 编译结果 1`] = `
"function anonymous(slice
) {
  /* aggregate function generated by dvt-aggregation@test
    收入: income = SUM(income)
    目标: target = SUM(target)
    code_1: code_1 = (target - income) / SUM(income)
    code_2: code_2 = MAX(target) - MIN(income)
    code_3: code_3 = MIN(target) - MAX(income)
    code_4: code_4 = code_2 + code_3
  */

  let r = {};

  r['income'] = 0;
  r['target'] = 0;
  r['safe_code_0'] = 0;
  r['safe_code_1'] = -Infinity;
  r['safe_code_4'] = -Infinity;
  r['safe_code_2'] = Infinity;
  r['safe_code_3'] = Infinity;

  for (const d of slice) {
    r['income'] += d['income'] || 0;
    r['target'] += d['target'] || 0;
    r['safe_code_0'] += d['income'] || 0;
    r['safe_code_1'] = Math.max(r['safe_code_1'], d['target']);
    r['safe_code_4'] = Math.max(r['safe_code_4'], d['income']);
    r['safe_code_2'] = Math.min(r['safe_code_2'], d['income']);
    r['safe_code_3'] = Math.min(r['safe_code_3'], d['target']);
  }

  r['code_1'] = ((r['target'] - r['income']) / r['safe_code_0']);
  r['code_2'] = (r['safe_code_1'] - r['safe_code_2']);
  r['code_3'] = (r['safe_code_3'] - r['safe_code_4']);
  r['code_4'] = (r['code_2'] + r['code_3']);

  return r;
}"
`;

exports[`简单的计算表达式 编译结果 1`] = `
"function anonymous(slice
) {
  /* aggregate function generated by dvt-aggregation@test
    收入: income = SUM(income)
    目标: target = SUM(target)
    达成率: act_rate = income / target
  */

  let r = {};

  r['income'] = 0;
  r['target'] = 0;

  for (const d of slice) {
    r['income'] += d['income'] || 0;
    r['target'] += d['target'] || 0;
  }

  r['act_rate'] = (r['income'] / r['target']);

  return r;
}"
`;
